internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -357,23 +378,68 @@ public VeniceWriter(
    */
   @Override
   public void close(boolean gracefulClose) {
-    try {
-      // If {@link #broadcastEndOfPush(Map)} was already called, the {@link #endAllSegments(boolean)}
-      // will not do anything (it's idempotent). Segments should not be ended if there are still data missing.
-      if (gracefulClose) {
-        endAllSegments(true);
+    synchronized (closeLock) {
+      if (isClosed) {
+        return;
       }
-      // DO NOT call the {@link #PubSubProducerAdapter.close(int) version from here.}
-      // For non-shared producer mode gracefulClose will flush the producer
+      long startTime = System.currentTimeMillis();
+      logger.info("Closing VeniceWriter for topic: {}", topicName);
+      try {
+        // If {@link #broadcastEndOfPush(Map)} was already called, the {@link #endAllSegments(boolean)}
+        // will not do anything (it's idempotent). Segments should not be ended if there are still data missing.
+        if (gracefulClose) {
+          endAllSegments(true);
+        }
+        // DO NOT call the {@link #PubSubProducerAdapter.close(int) version from here.}
+        // For non-shared producer mode gracefulClose will flush the producer
 
-      producerAdapter.close(topicName, closeTimeOut, gracefulClose);
-      OPEN_VENICE_WRITER_COUNT.decrementAndGet();
-    } catch (Exception e) {
-      logger.warn("Swallowed an exception while trying to close the VeniceWriter for topic: {}", topicName, e);
-      VENICE_WRITER_CLOSE_FAILED_COUNT.incrementAndGet();
+        producerAdapter.close(topicName, closeTimeOut, gracefulClose);
+        OPEN_VENICE_WRITER_COUNT.decrementAndGet();
+      } catch (Exception e) {
+        logger.warn("Swallowed an exception while trying to close the VeniceWriter for topic: {}", topicName, e);
+        VENICE_WRITER_CLOSE_FAILED_COUNT.incrementAndGet();
+      }
+      isClosed = true;
+      logger.info("Closed VeniceWriter for topic: {} in {} ms", topicName, System.currentTimeMillis() - startTime);
     }
   }
 
+  public CompletableFuture<VeniceResourceCloseResult> closeAsync(boolean gracefulClose) {
+    return CompletableFuture.supplyAsync(() -> {
+      synchronized (closeLock) {
```

Comment:
I think maybe I don't quite understand the code here, if all futures sync on the `closeLock`, would that make the close back to sequential close ?

Reply:
CloseLock is per VenieWriter hence there won't be contention on it during close.